Improve contribution guide (#404)
saleor/saleor-docs: docs/developer/community/contributing.mdx

These files should be updated by running `poetry export --without-hashes -f requirements.txt -o requirements.txt` and `poetry export --without-hashes -f requirements.txt -o requirements_dev.txt --dev`, respectively.

## File structure
We are using a standard Django structure - every app has its own directory, where you can find:
- migrations directory
- management directory
- models
- utils - that keeps some functions of general utility related to this module
- error codes - the definitions of errors that might appeared
- tests directory - that contains related tests,
- etc.

### API file structure
The `API` files are in `saleor/graphql/` directory. Every app has its own directory 
inside with:
- `schema.py` - with definitions of queries and mutations
- `sorters.py` - with definitions of sorters
- `filters.py` - with definitions of filters

> - `types.py` - with definitions of corresponding types

Reviewer: Also could be a directory like `mutations` for example in `products` or `core`.

- `enums.py` - with related enums

> - `dataloaders.py` - with data loaders for the given module.

Reviewer: Also could be a directory like `mutations`.

- mutations file or directory.

> - test directory.

Reviewer 2: ```suggestion
- tests directory.
```

We are aiming to have `mutations` directory in every module with a file per every mutation
how it's done in the `checkout` directory. See an example below:

```bash
.
└── saleor
    └── graphql
        └── checkout

>             ├── migrations

Reviewer 3: ```suggestion
            ├── mutations
```

>             ├── migrations

Reviewer: :police_officer: 
```suggestion
            ├── mutations
```

            │   └── checkout_create.py
            │   └── checkout_complete.py
            └── schema.py
```

### Tests file structure
We are keeping tests in `tests` directory. We are aiming to keep queries and mutations
in separate directories with a single file for every query and mutation. 
After joining it with the previous example, it would look like this:

```bash
.
└── saleor
    └── graphql
        └── checkout

>             ├── migrations

Reviewer 3: ```suggestion
            ├── mutations
```

            │   └── checkout_create.py
            │   └── checkout_complete.py
            ├── schema.py
            └── tests

>                 ├── migrations

Reviewer 3: ```suggestion
                ├── mutations
```

                │   ├── test_checkout_complete.py
                │   └── test_checkout_create.py
                └── queries

>                     └── test_checkout.py

Reviewer 4: Maybe we could add here:
```
queries:
 - test_checkout.py
 - test_checkouts.py
 ```
 Just to emphasize, we want to have  separate files for `query checkout(id:<id>)` and for `query checkouts(first:10)`

```

## Testing

> Testing are really important part of every project.

Reviewer 3: Grammarly feedback:
```suggestion
Testing is a really important part of every project.
```

The test file structure was introduced in the [tests file structure](contributing#tests-file-structure) part.
The main rule is not overloading test files. Always smaller files are preferable over big ones
where lots of logic is tested and it's hard to extend.
In the case of testing `API` we would like to split all tests into `mutations` and `queries`
section and test every query and mutation in a separate file.

### How to run tests?
To run tests just enter `pytest` in your terminal.
```bash
pytest
```

To speed time testing time, we recommend using the `reuse-db` flag.
```bash
pytest --reuse-db
```

### How to run particular tests?
As running all tests is quite time-consuming, sometimes you would like to run
only tests from one dictionary or even just a particular test.
You can use for that the following commands. In the case of a particular directory or file, 
just provide the path after the `pytest` command, like:
```bash
pytest --reuse-db saleor/graphql/app/tests
```

If you would like to run a particular test, you need to provide the path to the file
where the test is, and the file name after the `::` sign. In the case of running a single test

> it's also worth using `-n0` flag to run the test only in one thread. It will significantly decrease time.

Reviewer 3: Is it a typo or should it be `-n 0`? I always use `-n 0`.

Author: I'm using `-n0`, so I suppose there is no difference 😄

```

### Using pdb when testing
If you would like to use `pdb` in code when running a test, you need to use a few flags
`-n0` to one tests in a single thread, `-s` to disable capturing standard output by pytest, and
`--allow-hosts` with default port, as we disabled sockets by default.
So the previous example will look like this:

```bash
pytest --reuse-db saleor/graphql/app/tests/mutations/test_app_create.py::test_app_create_mutation -n0 -s --allow-hosts=127.0.0.1
```

### Recording cassettes
Some of our tests using `VCR.py` cassettes to recording request and response
from external API. To record one you need to use `vcr-record` flag and specified
`allow-hosts`:
```bash
pytest --vcr-record=once saleor/app/tests/test_app_commands.py --allow-hosts=127.0.0.1
```

### Writing benchmark tests
The benchmark tests allow us to keep track of a number of database queries 
for mutations and queries. Tests should be added for every new
mutation or query for objects.
The benchmark tests are in `benchmark` directory, every test must be marked with 
`pytest.mark.django_db` and `pytest.mark.count_queries(autouse=False)` decorators

> and must have `count_queries` argument.

Reviewer 4: Would be good to provide link to the package that we're using here: https://pypi.org/project/pytest-django-queries/

```python
@pytest.mark.django_db
@pytest.mark.count_queries(autouse=False)
def test_apps_for_federation_query_count(
    staff_api_client,
    permission_manage_apps,
    django_assert_num_queries,
    count_queries,
):
    ...
```

To check the number of queries, just run the benchmark test, and after that call following command in your terminal:
```bash
django-queries show
```
You will see the number of queries that were performed for each test
in every file.

### Given, when, then
It's recommended to use `given`, `when`, `then` to distinguish between different parts of the test.

> It significantly improves test readability and makes clear want you are testing at all.

Reviewer 2: ```suggestion
It significantly improves test readability and makes clear what you are testing at all.
```

Here we go with some tips and recommendation that limits the number of comments to your PR.
If you will follow those rules both sites will be happy. 

### Imports
Use relative imports.

### Models
The `UUID` is the preferred type of `PK` for the main models. Especially for models that
keeps sensitive data.

### Migrations
Try to combine multiple migrations into one, but remember to not mix changes on the database
with updating rows in migrations - in other words, operations that alter tables
and that which uses `RunPython` to run methods on existing data should be in separate files.

### Utility functions
Utility function specific for API should go to `graphql` directory,
otherwise to main module directory, usually to `utils.py` file.
When writing such a method try to find a descriptive name as possible.
Also, do not forget about the docstring, especially when the function is complicated.

### Searching
Right now, we are mainly using the `GinIndex` and `ilike` operator for searching but currently,
we are testing a new solution with the use of `SearchVector` and `SearchRank`.
You can find it in this PR [#9344](https://github.com/saleor/saleor/pull/9344).

### API

> Use `id` for mutation inputs instead od `model_name_id`

Reviewer 3: ```suggestion
Use `id` for mutation inputs instead of `model_name_id`
```

#### API field descriptions
Every mutation, mutation field, and type field should have a short, meaningful description,
ending with a dot. Also, we labeled every new field with info in which version it was
introduced, and when it's a new feature we added the preview feature label.
The `ADDED_IN_X` at the beginning of the description and `PREVIEW_FEATURE` label at the end.
All labels can be found in `saleor/graphql/core/descriptions.py`.

> When we would like to remove the API field, we marked it first as `DEPRECATED`.

Reviewer 2: ```suggestion
When we would like to remove the API field, we mark it first as `DEPRECATED`.
```

In a field definition, there is a dedicated argument for that: `deprecation_reason`,
but in the case of mutation arguments, we are using `DEPRECATED_IN_X_INPUT` label in the description.

> 

Reviewer 2: Double newline

and django `FilterSet` in dedicated `filters.py` file.
```python
class AppFilterInput(FilterInputObjectType):
    class Meta:
        filterset_class = AppFilter

class AppFilter(django_filters.FilterSet):
    type = EnumFilter(input_class=AppTypeEnum, method=filter_app_type)
    search = django_filters.CharFilter(method=filter_app_search)
    is_active = django_filters.BooleanFilter()

    class Meta:
        model = models.App
        fields = ["search", "is_active"]
```

Next, you need to create `CountableConnection` for given query:
```python
class AppCountableConnection(CountableConnection):
    class Meta:
        node = App
```

Then in the `schema.py` file, you need to change the corresponding query.
It must be a `FilterConnectionField` with created connection field, 
sorting input must be used in `sort_by` argument, and filter input in the `filter`
argument of the query:

```python
class AppQueries(graphene.ObjectType):
    apps = FilterConnectionField(
        AppCountableConnection,
        filter=AppFilterInput(description="Filtering options for apps."),
        sort_by=AppSortingInput(description="Sort apps."),
        description="List of the apps.",
    )
```

> #### How to handle breaking changes?

Reviewer 3: We should also list what changes we consider breaking.

Reviewer 3: Now I see we have it below. Maybe the these two sections should come together.

Author: These sections are together, the `What is considered as breaking changes?` (heading 3) is under the `changelog` section (heading 2). I decided to keep it like that to allow a quick finding of this part. It's shown in many part, under `changelog` section.
![image](https://user-images.githubusercontent.com/40886528/163345450-ba31b036-305c-45d8-bf61-fd36738e9235.png)

Reviewer 3: 👍

```python
class AppFilterInput(FilterInputObjectType):
    class Meta:
        filterset_class = AppFilter

class AppFilter(django_filters.FilterSet):
    type = EnumFilter(input_class=AppTypeEnum, method=filter_app_type)
    search = django_filters.CharFilter(method=filter_app_search)
    is_active = django_filters.BooleanFilter()

    class Meta:
        model = models.App
        fields = ["search", "is_active"]
```

Next, you need to create `CountableConnection` for given query:
```python
class AppCountableConnection(CountableConnection):
    class Meta:
        node = App
```

Then in the `schema.py` file, you need to change the corresponding query.
It must be a `FilterConnectionField` with created connection field, 
sorting input must be used in `sort_by` argument, and filter input in the `filter`
argument of the query:

```python
class AppQueries(graphene.ObjectType):
    apps = FilterConnectionField(
        AppCountableConnection,
        filter=AppFilterInput(description="Filtering options for apps."),
        sort_by=AppSortingInput(description="Sort apps."),
        description="List of the apps.",
    )
```

#### How to handle breaking changes?

> - provide compatibility layer by supporting two solutions for a specified time period and removing it in next major version

Reviewer 5: If it has been decided to use commas above(e.g. 39-44 lines) then please also add here.

:::

For more information and tips on how to write good commit messages, see the GitHub [guide](https://github.com/erlang/otp/wiki/writing-good-commit-messages).

## Pull requests
When you open the pull request remember to add a meaningful title and a good description.
Describe what is changing, and what is the reason for doing that or what problem it fixes.

> If it resolves some GitHub issue, please linked it. Wait for all actions to be performed,

Reviewer 2: ```suggestion
If it resolves some GitHub issue, please link it. Wait for all actions to be performed,
```

> and if all is green, request `saleor-core` group for review.

Reviewer 2: ain't that just `core` now?

## Changelog
We have a `CHANGELOG.md` file where we are keeping the info about the changes that were introduced.
The file contains separate parts for every release, the new items should be put under the section `Unreleased`.
This section in split into `Breaking changes` and `Other changes`.
The changelog entry should consist the name of the PR, the PR number and name of the author,
it may also contain an additional explanation. Here is an example:
```
- Add multichannel - #6242 by @exampleUser
```

### What is considered as breaking changes?
Here we go with the full list of changes that are considered by us as breaking:
- deleting field from graphql schema / renaming the field name
- deleting the field from webhook payload / changing the name of returned field
- changing signatures of plugins functions (PluginsManager) - breaking only for existing plugins
- adding new validation in mutations logic - it might breaks storefronts and apps
- changing of API behavior even if the schema doesn't change - e.g. the type of field hasn't changed but the requirements for value has

> Any change that is not specified there should go to `Other changes` section.

Author: AFAIR we shouldn't add a new changelog record in such a case, but @maarcingebala, could you confirm?

> Any change that is not specified there should go to `Other changes` section.

Reviewer 3: The changelog shouldn't reflect the history of building a feature, but it should provide information for other devs that the actual feature was added. If the feature wasn't released, and you believe the PR number of the fix is really worth mentioning, I would extend the existing changelog entry:
```
- my super great feature - #404, #999 by @korycins
```

If the fix is not so crucial, I wouldn't bother with mentioning it. If a feature was already released, you can add separate changelog entry just to mention a regular bugfix.

:::

For more information and tips on how to write good commit messages, see the GitHub [guide](https://github.com/erlang/otp/wiki/writing-good-commit-messages).

## Pull requests
When you open the pull request remember to add a meaningful title and a good description.
Describe what is changing, and what is the reason for doing that or what problem it fixes.
If it resolves some GitHub issue, please link it. Wait for all actions to be performed,
and if all is green, request `saleor/core` group for review.

> 

Reviewer 5: Please remove the redundant blank lines.
